Feature/FA-2-Forget-Password ✅

EslamMagdy12 opened the pull request:
This pull request implements the **Forget Password**, **Email Verification**, and **Reset Password** functionalities, integrating them seamlessly across the **data**, **domain**, and **presentation** layers. These enhancements ensure a smoother password recovery experience for users while improving code maintainability and structure. Additionally, minor optimizations and refactoring were done to enhance performance and maintain consistency across the project.

### Changes Included:

#### 🔹 Data Layer:
- Added API endpoints in `AuthRetrofitClient` for:
  - Sending a forget password request.
  - Verifying the reset code.
  - Resetting the password.
- Implemented **Auth Remote Data Source** to manage authentication-related API calls.
- Created request and response models for:
  - `forget_password`
  - `verify_reset_code`
  - `reset_password`
- Developed the **Auth Repository Implementation** to handle data fetching and processing.

#### 🔹 Domain Layer:
- Created new **use cases** for handling:
  - Forget password requests.
  - Reset code verification.
  - Password reset submission.
- Integrated **repository contracts** to ensure structured data handling.

#### 🔹 Presentation Layer:
- Developed dedicated screens for:
  - **Forget Password**
  - **Email Verification**
  - **Reset Password**
- Enhanced UI elements to provide a **more user-friendly experience**.
- Improved widget performance by using `const` where applicable.

#### 🔹 Localization & UI Enhancements:
- Updated Arabic & English translation files (`ar.json` & `en.json`) to include **new authentication-related strings**.
- Introduced **PIN input theme customizations** in `app_theme.dart`.

#### 🔹 General Improvements:
- **Refactored API Exception Handling:**
  - Renamed `app_exception.dart` to `api_exception.dart` for better clarity.
  - Updated all related imports accordingly.
  - Improved `_extractErrorMessage` in `ApiManager` to **handle API errors more effectively**.
- **Navigation Updates:**
  - Registered **Forget Password**, **Email Verification**, and **Reset Password** routes in `routes.dart`.

## 📌 Pull Request Checklist

### ✅ General
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] The PR title follows the required format: `type/FA-<issue-number>-TaskName`.
- [x] The branch name follows the required format: `feature/meaningful-name`, `fix/issue-description`, etc.

### 📂 Code Quality
- [x] All filenames are in `snake_case`.
- [x] All functions and variable names follow Dart naming conventions.
- [x] All new code is formatted with `dart format .`

### 🛠 CI/CD Validation
- [x] Linting passes with `dart analyze`.
- [x] The branch name is validated against the allowed patterns.
- [x] The PR title is validated against the required format.

### 🚀 Testing
- [ ] Unit tests are included for all new features and bug fixes.
- [ ] `flutter test --coverage` has been run and the results reviewed.
- [ ] The project builds successfully without errors.

### 📝 Additional Information
This update enhances the **password recovery process**, ensuring a structured and scalable implementation while improving user experience. Future updates will focus on adding **unit tests** to validate functionality and maintain reliability.

### 🔗 Related JIRA Ticket(s):
- [x] JIRA Issue: `FA-2`

### 📸 Screenshots (if applicable):
<div align="center">

|         Forget Password          |        Email Verification        |        Reset Password         |
|:---------------------:|:----------------------:|:----------------------:|
| <img src="https://github.com/user-attachments/assets/b0d40b43-e4cc-4f02-8a5c-7520fc83532c" width="300" alt="News"> | <img src="https://github.com/user-attachments/assets/38220ef3-d4ee-49f2-9cad-2aa2a81d43bc" width="300" alt="Favorite"> | <img src="https://github.com/user-attachments/assets/969ce0bf-cd11-4c60-87a3-f2017cf35a2c" width="300" alt="Details"> |

</div>

EslamMagdy12 committed:
feat(auth): implement forget and reset password data layer.

EslamMagdy12 committed:
feat(auth): implement forget and reset password domain layer.

EslamMagdy12 committed:
Merge branch 'refs/heads/developement' into Feature/FA-2-Forget-Password

EslamMagdy12 committed:
feat(auth): implement forget and reset password presentation layer using cubit as a state management.

EslamMagdy12 committed:
refactor(api): rename app_exception.dart to api_exception.dart and displayed error message returned from api.

EslamMagdy12 committed:
refactor(auth): update widgets to use const for better performance, remove unused imports, and remove relying on hardcoded colors and used the app_colors.

EslamMagdy12 committed:
refactor(auth): reformat and analyze codes using dart format and dart analyze.

ahmedelazab1220 commented on lib/features/auth/presentation/view/screens/email_verification.dart:
```diff
@@ -0,0 +1,65 @@
+    return Scaffold(
+      appBar: AppBar(title: Text(LocaleKeys.Password.tr())),
+      body: BlocProvider(
+        create: (context) => emailVerificationCubit,
+        child: BlocConsumer<EmailVerificationCubit, EmailVerificationState>(
+          listener: (context, state) {
```
Why use BlocConsumer , BlocListener is enough

EslamMagdy12 replied:
If I used BlocListener only I cant show error message "Invalid Input"

ahmedelazab1220 replied:
no you can make that.

if you need to rebuild something in ui you can use BlocBuilder on it instead of rebuild all screen.

EslamMagdy12 replied:
Got it, so BlocBuilder will be on the OTP only

ahmedelazab1220 replied:
yes

ahmedelazab1220 commented on assets/translations/ar.json:
you don't make any change so this is duplicate and you must remove it.

EslamMagdy12 replied:
There were just missing texts related to text form fields hint texts of reset password

ahmedelazab1220 replied:
ok add missing text only, not make duplicate by add text again

ahmedelazab1220 commented on assets/translations/en.json:
you don't make any change so this is duplicate and you must remove it.

EslamMagdy12 replied:
There were just missing texts related to text form fields hint texts of reset password

ahmedelazab1220 replied:
ok add missing text only, not make duplicate by add text again

ahmedelazab1220 commented on lib/core/theme/app_theme.dart:
```diff
@@ -124,4 +127,21 @@ class AppTheme {
       showSelectedLabels: true,
     ),
   );
+
+  static PinTheme defaultPinTheme = PinTheme(
+    width: 74.w,
+    height: 68.h,
+    textStyle: TextStyle(fontSize: 24.sp, fontWeight: FontWeight.w500),
+    decoration: BoxDecoration(
+      color: const Color(0xFFDFE7F7),
+      borderRadius: BorderRadius.circular(10).r,
+    ),
+  );
+
+  static PinTheme errorPinTheme = defaultPinTheme.copyWith(
+    decoration: BoxDecoration(
+      border: Border.all(color: const Color(0xFFCC1010)),
+      borderRadius: BorderRadius.circular(10).r,
+    ),
+  );
```
this is not place to set it and you can use textformfield instead of pin

use dart format(apply linting).

ahmedelazab1220 commented on lib/core/utils/datasource_excution/api_manager.dart:
```diff
@@ -110,7 +110,7 @@ class ApiManager {
   String _extractErrorMessage(dynamic data) {
     if (data is Map<String, dynamic>) {
-      return data['message']?.toString() ??
+      return data['error']?.toString() ??
```
why make this change!, api return error in "message"

EslamMagdy12 replied:
Actually Api returns error via "error" not "message"

ahmedelazab1220 replied:
api in this project return errors in "message" not error you can make like that

data['message']?.toString() ?? data['error']?.toString() ?? .....


to make suitable for differenet apis.

EslamMagdy12 replied:
This is response of Verifying Reset Code

{
    "error": "Reset code is invalid or has expired"
}

but anyway as you said made it work for both `message` and `error`

ahmedelazab1220 commented on lib/core/utils/datasource_excution/api_exception.dart:
don't change setup (by rename file) this is not good.

EslamMagdy12 replied:
Thats ok just changed its name because class name is ApiException

ahmedelazab1220 replied:
return to old name is better, because this file is responsiable for all exceptions and handle it.

ahmedelazab1220 commented on lib/features/auth/data/data_source/contract/auth_remote_data_source.dart:
```diff
@@ -0,0 +1,20 @@
+import '../../model/verify_reset_code/verify_reset_code_request_dto.dart';
+import '../../model/verify_reset_code/verify_reset_code_response_dto.dart';
+
+abstract interface class AuthRemoteDataSource {
```
remove interface keyword

EslamMagdy12 replied:
It's better to keep it interface so no one can extends it, so you must use keyword implements in other classes

ahmedelazab1220 replied:
abstract is enough

ali72-20 commented on lib/features/auth/data/model/forget_password/forget_password_request_dto.dart:
```diff
@@ -0,0 +1,19 @@
+@JsonSerializable()
+class ForgetPasswordRequestDto {
+  @JsonKey(name: "email")
+  final String? email;
```
avoid using nullable in request models 
all data is required

ali72-20 commented on lib/features/auth/data/model/reset_password/reset_password_request_dto.dart:
```diff
@@ -0,0 +1,21 @@
+part 'reset_password_request_dto.g.dart';
+
+@JsonSerializable()
+class ResetPasswordRequestDto {
```
avoid using nullable in request models

ali72-20 commented on lib/features/auth/data/repo_impl/auth_repo_impl.dart:
```diff
@@ -0,0 +1,55 @@
+  Future<Result<void>> forgetPassword(
+    ForgetPasswordRequestDto forgetPasswordRequestDto,
+  ) {
+    var response = _apiManager.execute(() async {
```
to can add return here 
return _apiManger.excure

ali72-20 commented on lib/features/auth/domain/usecase/verify_reset_code_usecase.dart:
```diff
@@ -0,0 +1,15 @@
+
+  VerifyResetCodeUsecase(this._authRepo);
+
+  Future<Result<void>> call(VerifyResetCodeRequestDto verifyResetCodeRequest) {
```
don't forget async and await

ali72-20 reviewed (changes requested):
take care about this small things 

Greate work eslam 
دي اقل حاجة

ahmedelazab1220 commented:
name of branch must be feature/fa-2-forget-password

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/reset_password_state.dart:
```diff
@@ -0,0 +1,13 @@
+  final String message;
+
+  ResetPasswordFailure(this.message);
+}
```
use base state and extends equtable to improve performance and work on standard we agree on it.

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/forget_password_state.dart:
```diff
@@ -0,0 +1,13 @@
+  final String message;
+
+  ForgetPasswordFailure({required this.message});
+}
```
use base state and extends equtable to improve performance and work on standard we agree on it.

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/email_verification_state.dart:
```diff
@@ -0,0 +1,23 @@
+  final String message;
+
+  ResendEmailVerificationFailure(this.message);
+}
```
use base state and extends equtable to improve performance and work on standard we agree on it.

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/email_verification_cubit.dart:
```diff
@@ -0,0 +1,53 @@
+class EmailVerificationCubit extends Cubit<EmailVerificationState> {
+  final VerifyResetCodeUsecase verifyResetCodeUsecase;
+  final ForgetPasswordUsecase forgetPasswordUsecase;
+  final Validator validator = Validator();
```
use constructor injection instead of make that

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/forget_password_cubit.dart:
```diff
@@ -0,0 +1,36 @@
+  ForgetPasswordCubit(this.forgetPasswordUsecase)
+    : super(ForgetPasswordInitial());
+
+  TextEditingController emailController = TextEditingController();
+
```
use dispose function to avoid memory leaks

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/email_verification_cubit.dart:
```diff
@@ -0,0 +1,53 @@
+    this.forgetPasswordUsecase,
+  ) : super(EmailVerificationInitial());
+
+  final TextEditingController pinController = TextEditingController();
```
use dispose function to avoid memory leaks

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,72 @@
+                    const SizedBox(width: 4),
+                    Text(
+                      errorText!,
+                      style: TextStyle(color: AppColors.red, fontSize: 13.sp),
```
13.sp not work on all screen.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,72 @@
+            },
+          ),
+        ),
+        SizedBox(height: 24.h),
```
this fixed on all screens and don't need to use responsive.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,72 @@
+                    Icon(
+                      Icons.error_outline,
+                      color: AppColors.red,
+                      size: 16.sp,
```
16.sp not work on all screen.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,76 @@
+              ),
+              SizedBox(height: 48.h),
+              SizedBox(
+                width: double.infinity,
+                height: 48.h,
```
48.h this not good on all screen remove it please and remove double.infinity and make column streatch is better.

EslamMagdy12 committed:
refactor(auth): Refactor Forget Password Feature
- All fields are required in request models.
- Added some missed async and await.
- Removed Pintheme from AppTheme.
- Removed BlocConsumer and used BlocListener Instead.
- Renamed again ApiException to AppException.
- Followed the same Architecture so moved Auth Remote DataSource Impl outside implementation folder

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,76 @@
+                  hintText: LocaleKeys.ConfirmYourPassword.tr(),
+                ),
+                validator:
+                    (value) => context
+                        .read<ResetPasswordCubit>()
```
don't make that , make shared variable on file like

var viewModel = context.read<ResetPasswordCubit>();

then access it in any place in file this is better.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,76 @@
+              SizedBox(height: 24.h),
+              TextFormField(
+                controller:
+                    context
+                        .read<ResetPasswordCubit>()
+                        .confirmPasswordController,
```
don't make that , make shared variable on file like

var viewModel = context.read<ResetPasswordCubit>();

then access it in any place in file this is better.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,76 @@
+          child: Column(
+            children: [
+              TextFormField(
+                controller:
+                    context.read<ResetPasswordCubit>().passwordController,
```
don't make that , make shared variable on file like

var viewModel = context.read<ResetPasswordCubit>();

then access it in any place in file this is better.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,76 @@
+    return Column(
+      children: [
+        Form(
+          key: context.read<ResetPasswordCubit>().formKey,
```
don't make that , make shared variable on file like

var viewModel = context.read<ResetPasswordCubit>();

then access it in any place in file this is better.

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/email_verification_cubit.dart:
use mvi pattern

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/forget_password_cubit.dart:
use mvi pattern

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/reset_password_cubit.dart:
use mvi pattern 

use dispose function to avoid memory leaks

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/reset_password_cubit.dart:
```diff
@@ -0,0 +1,40 @@
+@injectable
+class ResetPasswordCubit extends Cubit<ResetPasswordState> {
+  final ResetPasswordUsecase _resetPasswordUsecase;
+  final Validator validator = Validator();
```
use constructor injectation instead of make that.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,72 @@
+  final EmailVerificationState state;
+
+  const VerificationCodeInput({super.key, required this.state});
+
```
this is not good pass state from file to another 

you can use read cubit and access state.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/forget_password_form.dart:
```diff
@@ -0,0 +1,47 @@
+          ),
+          SizedBox(height: 48.h),
+          SizedBox(
+            width: double.infinity,
+            height: 48.h,
```
48.h this not good on all screen remove it please and remove double.infinity and make column streatch is better.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/forget_password_form.dart:
```diff
@@ -0,0 +1,47 @@
+      child: Column(
+        children: [
+          TextFormField(
+            controller: context.read<ForgetPasswordCubit>().emailController,
```
don't make that , make shared variable on file like

var viewModel = context.read();

then access it in any place in file this is better.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/forget_password_form.dart:
```diff
@@ -0,0 +1,47 @@
+              prefixIcon: const Icon(Icons.email),
+            ),
+            validator:
+                (value) => context
+                    .read<ForgetPasswordCubit>()
```
don't make that , make shared variable on file like

var viewModel = context.read();

then access it in any place in file this is better.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/forget_password_form.dart:
```diff
@@ -0,0 +1,47 @@
+  @override
+  Widget build(BuildContext context) {
+    return Form(
+      key: context.read<ForgetPasswordCubit>().formKey,
```
don't make that , make shared variable on file like

var viewModel = context.read();

then access it in any place in file this is better.

ahmedelazab1220 commented on lib/features/auth/presentation/view/screens/reset_password.dart:
```diff
@@ -0,0 +1,58 @@
+                children: [
+                  ForgetPasswordHeader(
+                    title: LocaleKeys.ResetPassword.tr(),
+                    subtitle: LocaleKeys.PasswordRequirements.tr(),
```
this not you make custom widget for more than one screen but name of class is not good you can rename to 

HeaderWidget this is better because is custom for more than one screen.

ahmedelazab1220 commented on lib/features/auth/presentation/view/screens/reset_password.dart:
```diff
@@ -0,0 +1,58 @@
+          },
+          builder: (context, state) {
+            return Padding(
+              padding: const EdgeInsets.all(16).w,
```
incorrect use of screenUtil 

and set padding with responsive values is not good in this case.

ahmedelazab1220 commented on lib/features/auth/presentation/view/screens/forget_password.dart:
```diff
@@ -0,0 +1,66 @@
+                    subtitle:
+                        LocaleKeys
+                            .PleaseEnterYourEmailAssociatedWithToYourAccount.tr(),
+                  ),
```
this not you make custom widget for more than one screen but name of class is not good you can rename to

HeaderWidget this is better because is custom for more than one screen.

ahmedelazab1220 commented on lib/features/auth/presentation/view/screens/email_verification.dart:
```diff
@@ -0,0 +1,65 @@
+                    subtitle:
+                        LocaleKeys
+                            .PleaseEnterYourCodeThatSendToYourEmailAddress.tr(),
+                  ),
```
this not you make custom widget for more than one screen but name of class is not good you can rename to

HeaderWidget this is better because is custom for more than one screen.

ahmedelazab1220 commented on lib/features/auth/presentation/view/screens/forget_password.dart:
```diff
@@ -0,0 +1,66 @@
+          },
+          builder: (context, state) {
+            return Padding(
+              padding: const EdgeInsets.all(16).w,
```
incorrect use of screenUtil

and set padding with responsive values is not good in this case.

ahmedelazab1220 commented on lib/features/auth/presentation/view/screens/forget_password.dart:
```diff
@@ -0,0 +1,66 @@
+import '../widgets/forget_password_form.dart';
+import '../widgets/forget_password_header.dart';
+
+class ForgetPassword extends StatelessWidget {
+  ForgetPassword({super.key});
+
+  final ForgetPasswordCubit forgetPasswordCubit = getIt<ForgetPasswordCubit>();
```
statelessWidget must immutable and constructor must const this is not good and affect on performance use stateful widget is better of make that and set this line on it 


final ForgetPasswordCubit forgetPasswordCubit = getIt<ForgetPasswordCubit>();

ahmedelazab1220 commented on lib/features/auth/presentation/view/screens/email_verification.dart:
```diff
@@ -0,0 +1,65 @@
+          },
+          builder: (context, state) {
+            return Padding(
+              padding: const EdgeInsets.all(16.0).w,
```
incorrect use of screenUtil

and set padding with responsive values is not good in this case.

EslamMagdy12 commented:
When I changed branch name to pass tests it required to make new pull request
so here is my new one [Pull Request #11](https://github.com/ahmedelazab1220/ecommerce_flower_app/pull/11)
